c++/src/Reader.cc

```diff
@@ -1186,41 +1186,46 @@ namespace orc {
                                            uint64_t currentRowInStripe,
                                            uint64_t rowsInCurrentStripe,
                                            uint64_t rowIndexStride,
-                                           const std::vector<bool>& includedRowGroups) {
+                                           const std::vector<uint64_t>& nextSkippedRows) {
     // In case of PPD, batch size should be aware of row group boundaries. If only a subset of row
     // groups are selected then marker position is set to the end of range (subset of row groups
     // within stripe).
     uint64_t endRowInStripe = rowsInCurrentStripe;
-    if (!includedRowGroups.empty()) {
-      endRowInStripe = currentRowInStripe;
-      uint32_t rg = static_cast<uint32_t>(currentRowInStripe / rowIndexStride);
-      for (; rg < includedRowGroups.size(); ++rg) {
-        if (!includedRowGroups[rg]) {
-          break;
-        } else {
-          endRowInStripe = std::min(rowsInCurrentStripe, (rg + 1) * rowIndexStride);
-        }
-      }
+    uint64_t groupsInStripe = nextSkippedRows.size();
+    if (groupsInStripe > 0) {
+      auto rg = static_cast<uint32_t>(currentRowInStripe / rowIndexStride);
+      if (rg >= groupsInStripe) return 0;
+      uint64_t nextSkippedRow = nextSkippedRows[rg];
+      if (nextSkippedRow == 0) return 0;
+      endRowInStripe = nextSkippedRow;
     }
     return std::min(requestedSize, endRowInStripe - currentRowInStripe);
   }
 
   uint64_t RowReaderImpl::advanceToNextRowGroup(uint64_t currentRowInStripe,
                                                 uint64_t rowsInCurrentStripe,
                                                 uint64_t rowIndexStride,
-                                                const std::vector<bool>& includedRowGroups) {
-    if (!includedRowGroups.empty()) {
-      uint32_t rg = static_cast<uint32_t>(currentRowInStripe / rowIndexStride);
-      for (; rg < includedRowGroups.size(); ++rg) {
-        if (includedRowGroups[rg]) {
-          return currentRowInStripe;
-        } else {
-          // advance to start of next row group
-          currentRowInStripe = (rg + 1) * rowIndexStride;
-        }
-      }
-    }
-    return std::min(currentRowInStripe, rowsInCurrentStripe);
+                                                const std::vector<uint64_t>& nextSkippedRows) {
+    auto groupsInStripe = nextSkippedRows.size();
+    if (groupsInStripe == 0) {
+      // No PPD, keeps using the current row in stripe
+      return std::min(currentRowInStripe, rowsInCurrentStripe);
+    }
+    auto rg = static_cast<uint32_t>(currentRowInStripe / rowIndexStride);
+    if (rg >= groupsInStripe) {
+      // Points to the end of the stripe
+      return rowsInCurrentStripe;
+    }
+    if (nextSkippedRows[rg] != 0) {
+      // Current row group is selected
+      return currentRowInStripe;
+    }
```

Comment:
Does Line 1219-1221 have redundancy with the following logic(Line 1224-1227)? It seems that deleting Line 1219-1221 does not affect the results.

Reply:
I think we need this. This is the only case that returns `currentRowInStripe`. If we delete this, we will return `rg * rowIndexStride` which is a different result. Or did I miss something?

Reply:
You're right. When the nextSkippedRows[rg] is not zero, it should return currentRowInStripe.